oxidation/libparsec/crates/crypto/src/rustcrypto/sequester.rs

```diff
@@ -0,0 +1,314 @@
+        EncodePublicKey,
+    },
+    pss::{Signature, SigningKey, VerifyingKey},
+    signature::RandomizedSigner,
+    PaddingScheme, PublicKey, RsaPrivateKey, RsaPublicKey,
+};
+use serde::{Deserialize, Serialize};
+use serde_bytes::Bytes;
+use sha1::Sha1;
+use sha2::Sha256;
+
+use crate::{CryptoError, CryptoResult, EnforceDeserialize, EnforceSerialize, SecretKey};
+
+/*
+ * PrivateKey
+ */
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct SequesterPrivateKeyDer(RsaPrivateKey);
+
+crate::impl_key_debug!(SequesterPrivateKeyDer);
+
+impl Default for SequesterPrivateKeyDer {
+    fn default() -> Self {
+        let mut rng = rand_08::thread_rng();
+        Self(RsaPrivateKey::new(&mut rng, Self::SIZE_IN_BITS).expect("Unreachable"))
+    }
+}
+
+impl TryFrom<&[u8]> for SequesterPrivateKeyDer {
+    type Error = CryptoError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        RsaPrivateKey::from_pkcs8_der(bytes)
+            .map(Self)
+            .map_err(|err| CryptoError::SequesterPrivateKeyDer(err.to_string()))
+    }
+}
+
+impl EnforceDeserialize for SequesterPrivateKeyDer {
+    const SIZE_IN_BITS: usize = Self::SIZE_IN_BITS;
+    const ALGORITHM: &'static [u8] = b"RSAES-OAEP-XSALSA20-POLY1305";
+}
+
+impl SequesterPrivateKeyDer {
+    #[cfg(test)]
+    pub const SIZE_IN_BITS: usize = 1024;
+    #[cfg(not(test))]
+    pub const SIZE_IN_BITS: usize = 4096;
+
+    pub fn dump(&self) -> Zeroizing<Vec<u8>> {
+        self.0.to_pkcs8_der().expect("Unreachable").to_bytes()
+    }
+
+    pub fn dump_pem(&self) -> Zeroizing<String> {
+        self.0
+            .to_pkcs8_pem(rsa::pkcs8::LineEnding::default())
+            .expect("Unreachable")
+    }
+
+    pub fn load_pem(s: &str) -> CryptoResult<Self> {
+        RsaPrivateKey::from_pkcs8_pem(s)
+            .map(Self)
+            .map_err(|err| CryptoError::SequesterPrivateKeyDer(err.to_string()))
+    }
+
+    pub fn decrypt(&self, data: &[u8]) -> CryptoResult<Vec<u8>> {
+        let (cipherkey, ciphertext) = Self::deserialize(data)?;
+        let padding = PaddingScheme::new_oaep::<Sha1>();
+
+        let clearkey = SecretKey::try_from(
+            &self
+                .0
+                .decrypt(padding, cipherkey)
+                .map_err(|_| CryptoError::Decryption)?[..],
+        )?;
+
+        clearkey.decrypt(ciphertext)
+    }
+}
+
+/*
+ * PublicKey
+ */
+
+#[derive(Clone, PartialEq, Eq, Deserialize)]
+#[serde(try_from = "&Bytes")]
+pub struct SequesterPublicKeyDer(RsaPublicKey);
+
+crate::impl_key_debug!(SequesterPublicKeyDer);
+
+impl From<SequesterPrivateKeyDer> for SequesterPublicKeyDer {
+    fn from(priv_key: SequesterPrivateKeyDer) -> Self {
+        Self(RsaPublicKey::from(priv_key.0))
+    }
+}
+
+impl EnforceSerialize for SequesterPublicKeyDer {
+    const SIZE_IN_BITS: usize = Self::SIZE_IN_BITS;
+    const ALGORITHM: &'static [u8] = b"RSAES-OAEP-XSALSA20-POLY1305";
+}
+
+impl SequesterPublicKeyDer {
+    #[cfg(test)]
```

Comment:
A comment is needed to explain why this is needed

Reply:
And I'm not a huge fan of this given it means we break make the test incompatible with the regular build: if we want to load a key obtain from the production app into the test this will fail in an unexpected way :/

Reply:
On top of that what happen if we use `load_pem` on a 3072bits long key ?

We need to support arbitrary key size (given the key generation is not handled by Parsec).

So in the end my guess is we should do:
- add the size of the key as a parameter to the `generate()` method
- support arbitrary size key
- add tests to make sure we support arbitrary size key (load/dump/encrypt/decrypt/sign/verify/serialization/deserialization)

Reply:
Done